dfdaemon/proxy/proxy.go

```diff
@@ -116,3 +227,59 @@ func copyHeader(dst, src http.Header) {
 		}
 	}
 }
+
+var okHeader = []byte("HTTP/1.1 200 OK\r\n\r\n")
+
+// handshake hijacks w's underlying net.Conn, responds to the CONNECT request
+// and manually performs the TLS handshake.
+func handshake(w http.ResponseWriter, config *tls.Config) (net.Conn, error) {
+	raw, _, err := w.(http.Hijacker).Hijack()
+	if err != nil {
+		http.Error(w, "no upstream", http.StatusServiceUnavailable)
+		return nil, err
+	}
+	if _, err = raw.Write(okHeader); err != nil {
+		raw.Close()
+		return nil, err
+	}
+	conn := tls.Server(raw, config)
+	if err = conn.Handshake(); err != nil {
+		conn.Close()
+		raw.Close()
+		return nil, err
+	}
+	return conn, nil
+}
+
+// A singleUseListener implements a net.Listener that returns the net.Conn specified
+// in c for the first Accept call, and return errors for the subsequent calls.
+type singleUseListener struct {
+	c net.Conn
+}
+
+func (l *singleUseListener) Accept() (net.Conn, error) {
+	if l.c == nil {
+		return nil, errors.New("closed")
+	}
+	c := l.c
+	l.c = nil
+	return c, nil
+}
+
+func (l *singleUseListener) Close() error { return nil }
+
+func (l *singleUseListener) Addr() net.Addr { return l.c.LocalAddr() }
+
+// A customCloseConn implements net.Conn and calls f before closing the underlying net.Conn
+type customCloseConn struct {
```

Comment:
Whether need to thread safety of `customCloseConn`?

Reply:
We're directly passing `&singleUseListener{&customCloseConn{sConn, wg.Done}}` to `http.Serve`, so no other goroutines in dfdaemon will access it. `singleUseListener` also ensures that the `customCloseConn` will only be returned once. I also took a quick look at the source code of `http.Serve`, since we're only handling a single connection at a time, nothing seems to suggest that `Close` will be called multiple times concurrently.

Is there any particular reason to make sure it's thread-safe?

Reply:
SGTM